### PR TITLE
New genesis structure changes

### PR DESCRIPTION
**Pull request by iulianpascalau**
refactor to match the modifications done in feat/genesis2

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -6,22 +6,36 @@ github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ
...
 github.com/ElrondNetwork/arwen-wasm-vm v0.3.10/go.mod h1:hF6eHBO8xL8LXPxcnV5Ygxv91I77dT/m1n7cDIKbjYE=
+github.com/ElrondNetwork/arwen-wasm-vm v0.3.15/go.mod h1:dVwR0ZGVOlr7BH9YbacSBWQ3CSASSp5BQU1tTiD/dOA=
 github.com/ElrondNetwork/big-int-util v0.0.5/go.mod h1:96viBvoTXLjZOhEvE0D+QnAwg1IJLPAK6GVHMbC7Aw4=
 github.com/ElrondNetwork/concurrent-map v0.1.2 h1:mr2sVF2IPDsJO8DNGzCUiNQOJcadHuIRVZn+QFnCBlE=
 github.com/ElrondNetwork/concurrent-map v0.1.2/go.mod h1:3XwSwn4JHI0lrKxWLZvtp53Emr8BXYTmNQGwcukHJEE=
 github.com/ElrondNetwork/elrond-go v1.0.90-0.20200313111119-11611acd9885 h1:4t02I4tjSVPTRzW+a7HijylKCtPsf/FQX0GqGnOaDy8=
 github.com/ElrondNetwork/elrond-go v1.0.90-0.20200313111119-11611acd9885/go.mod h1:U4fe6UTpGX4dL1ojL8cRNtlPvhfbmZiUxt661js10jU=
 github.com/ElrondNetwork/elrond-go v1.0.98-0.20200330193748-5f476a20792a h1:WYlpCj0vOoDH0JUMo4c2cXPwtIkHSE1bFGxouYj5xfY=
 github.com/ElrondNetwork/elrond-go v1.0.98-0.20200330193748-5f476a20792a/go.mod h1:+oV7zq2/19S9xnqScORoW0tYgYNrP2WnuheMaIWRzQw=
+github.com/ElrondNetwork/elrond-go v1.0.98-0.20200424100749-774ef1175544 h1:gN84cueoK92Kf6H2LJqTxb0vbAGHHQcwR4szuN7GLTw=
```
Comment (sasurobert) [inline go.sum:17]: go mod tidy

File: cmd/filegen/main.go [hunk 2]
```diff
@@ -377,43 +374,86 @@ func generateFiles(ctx *cli.Context) error {
...
-	for i := 0; i < totalAddressesWithBalances; i++ {
-		pkHex, skHex, err = getIdentifierAndPrivateKey(balancesKeyGenerator, pubKeyConverterTxs)
+	//initialTotalBalance = totalSupply - (numValidators * nodePriceValue)
+	initialTotalBalance := big.NewInt(0).Set(totalSupplyValue)
+	staked := big.NewInt(0).Mul(big.NewInt(int64(numValidators)), nodePriceValue)
+	initialTotalBalance.Sub(initialTotalBalance, staked)
+
+	// initialNodeBalance = initialTotalBalance / (totalAddressesWithBalances + numOfAdditionalAccounts)
+	initialNodeBalance := big.NewInt(0).Set(initialTotalBalance)
+	initialNodeBalance.Div(initialNodeBalance,
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/filegen/main.go:392]: of course. For BoN we have the elrond-go-tools variant, please check here: https://github.com/ElrondNetwork/elrond-go-tools/blob/genesis-refactor/cmd/csvJsonConverter/main.go
As for mainnet, we do not have (yet) the application that creates the genesis config file

File: cmd/filegen/main.go [hunk 3]
```diff
@@ -434,91 +474,178 @@ func generateFiles(ctx *cli.Context) error {
...
+			nodes.InitialNodes = append(nodes.InitialNodes, node)
+		} else {
+			ia.StakingValue = big.NewInt(0)
+			ia.Supply = big.NewInt(0).Set(ia.Balance)
+			ia.Delegation.Address = ""
+			ia.Delegation.Value = big.NewInt(0)
 		}
 	}
 
 	for i := 0; i < numOfAdditionalAccounts; i++ {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/filegen/main.go:495]: done

File: go.sum (hunk 1 above)
Comment (iulianpascalau) [reply go.sum:17]: done

File: cmd/filegen/main.go [hunk 4]
```diff
@@ -377,148 +373,347 @@ func generateFiles(ctx *cli.Context) error {
...
+			txgenAccounts[shardID] = append(txgenAccounts[shardID], txGenAccount)
 		}
+
+		genesisList = append(genesisList, ia)
 	}
 
-	genesisBuff, err := json.MarshalIndent(genesis, "", "  ")
+	//take the remainder and set it on the first node
+	// initialBalance = initialTotalBalance - (totalAddressesWithBalances + numOfAdditionalAccounts - 1) * initialNodeBalance
+	iaFirst := genesisList[0]
```
Comment (bogdan-rosianu) [inline cmd/filegen/main.go:481]: rename to `firstInitialAccount`

File: README.md [hunk 5]
```diff
@@ -15,8 +15,11 @@ Run in  %project_folder%/cmd/filegen folder the following command to build a fil
  
  ```
  $ go build
-$ ./filegen -mint-value 1000000000000000000000000000 -num-of-shards 5 -num-of-nodes-in-each-shard 21 -consensus-group-size 15 -num-of-observers-in-each-shard 1 -num-of-metachain-nodes 21 -metachain-consensus-group-size 15 -num-of-observers-in-metachain 1 -chain-id testnet -hysteresis 0.0
+$ ./filegen -node-price 500000000000000000000000 -total-supply 20000000000000000000000000000 -num-of-shards 5 -num-of-nodes-in-each-shard 21 -consensus-group-size 15 -num-of-observers-in-each-shard 1 -num-of-metachain-nodes 21 -metachain-consensus-group-size 15 -num-of-observers-in-metachain 1 -chain-id testnet -hysteresis 0.0
  ```
+
+There is an optional flag called `-stake-type` that can have one of the 2 values: `direct` and `delegated`. If the `delegated` 
+option is selected, the initial stking will be done through the delegation SC.
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply README.md:22]: done

File: cmd/filegen/main.go [hunk 6]
```diff
@@ -116,17 +124,32 @@ VERSION:
 		Name:  "txgen",
 		Usage: "If set, will generate the accounts.json file needed for txgen",
 	}
+	stakeType = cli.StringFlag{
+		Name: "stake-type",
+		Usage: "defines the 2 possible way to stake the nodes: 'direct' as in direct staking " +
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/filegen/main.go:129]: done

File: cmd/filegen/main.go [hunk 7]
```diff
@@ -264,100 +284,76 @@ func generateFiles(ctx *cli.Context) error {
...
-			fmt.Println(err.Error())
-		}
-		err = genesisFile.Close()
-		if err != nil {
-			fmt.Println(err.Error())
-		}
-		err = nodesFile.Close()
-		if err != nil {
-			fmt.Println(err.Error())
+		closeWithLog(walletKeyFile)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/filegen/main.go:312]: done

**Pull request merged**